feat: support bug detection related to isolation level

Tang-Kin-Long opened the pull request:
feat: support bug detection related to isolation level

resolve the issue: [bug detection related to isolation level](https://github.com/Tencent/3TS/issues/43)

Modify parts of code in [Coo consistency check](https://github.com/Tencent/3TS/pull/50/commits) to realize the error detection result output of related isolation level.

Reference the [test data](https://github.com/Tencent/3TS/pull/50/commits/f4909ca6c6a4add3778731851d7be63bc65ef0b2) as well.

Sample Output:
--------file:rr_s_s_avoid--------
Transaction 1:-----repeatable-read-----
  Error Detection: No error in this isolation level.
Transaction 2:-----serializable-----
  Edge(begin_time=4, type=ICR, out=3)
  Error Detection: Phantom-Read
Transaction 3:-----serializable-----
  Error Detection: No error in this isolation level.
---------------------------------#

Tang-Kin-Long committed:
输出隔离级别相关的异常检测结果

tencent-adm commented:
[![CLA assistant check](https://cla.opensource.tencent.com/pull/badge/not_signed)](https://cla.opensource.tencent.com/Tencent/3TS?pullRequest=57) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/3TS?pullRequest=57) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/3TS?pullRequest=57) it.</sub>

hldgs commented:
Hello, I am one of the code reviewers. Can you translate your **pr** title, description, commit messages to English? And we recommend using this [format](https://www.conventionalcommits.org/zh-hans/v1.0.0/#%e7%ba%a6%e5%ae%9a%e5%bc%8f%e6%8f%90%e4%ba%a4%e8%a7%84%e8%8c%83) to title your PR.

hldgs commented:
you can use "feat: support bug detection related to isolation level", and please add the issues you resolved in description.

Tang-Kin-Long commented:
> you can use "feat: support bug detection related to isolation level", and please add the issues you resolved in description.

Okay!

hldgs reviewed (changes requested):
if any questions, reply me directly

Tang-Kin-Long committed:
style: maintain function description of mda_detect.py

style: Update contents in docstring to maintain function description of mda_detect.py.

Tang-Kin-Long commented:
> if any questions, reply me directly

I have made the expected modification, please [check](https://github.com/Tencent/3TS/pull/57/commits/8b78e88e0e08b4523028bf64f638da3a5c08f43b), thank you.

hldgs commented on src/dbtest/src/mda_detect.py:
```diff
@@ -382,53 +599,108 @@ def insert_record(op_time, txn_num, txn, data_op_list):
         return error_message
     if op_time == -1 or txn_num == -1:
         return error_message
-    if query.find("BEGIN") != -1: # TODO: Need a related interface, I assume that it is read from the do_test_list file.:
```
if necessary, todo can be kept

Tang-Kin-Long replied:
I see. I think it is not so necessary. Perhaps I can put it into the function description above.

hldgs commented:
do you test your commit on every dbmss?

Tang-Kin-Long commented:
> do you test your commit on all dbmss?

No, I haven't installed Docker and configured the test environment. I am trying to do that.

hldgs commented on src/dbtest/src/mda_detect.py:
```diff
@@ -33,6 +36,62 @@ class Txn:
     def __init__(self):
         self.begin_ts = -1
         self.end_ts = 99999999999999999999
+        self.isolation = "serializable"
```
ser rr ru rc are not surpported by all dbms, some dbms such as mongodb dont support these 4 levels. you can firstly limit the scope of your feature on mysql, and we can extend to other dbms one by one

Tang-Kin-Long replied:
OK. In mysql, ser rr ru rc are all supported. I want to confirm whether my task is to run test samples on mysql, determine which bug occurs (or directly use test result in src/dbtest), and compare if this bug is the same as the mda_detect.py program predicts?

Tang-Kin-Long replied:
I have commited the test program. It passes test cases in MySQL at "serializable" isolation level. If you can accept this test procedure, I will continue to add functions for other isolation level and dbms. Thank you!

Tang-Kin-Long replied:
I have commited the improved [test program](https://github.com/Tencent/3TS/pull/57/commits/7965abb84e58379d8f2a6703d5b2130bc31a177a). Now, it passes test cases in MySQL at "repeatable-read" and "serializable" isolation levels. I will continue to add functions for other isolation level and dbms. Thank you!

Tang-Kin-Long committed:
test: pass test case in Mysql at "serializable" isolation level

This program adds a determination function to detect test cases in MySQL environment. For now, only "serializable" isolation level is supported. Other isolation levels and dbms environments will be added later.
The output result passes almost all test cases, except "iat_dda_write_skew_predicate_based-overdraft_protection", because its variable declaration is quite different.

Tang-Kin-Long committed:
test: pass test case in Mysql at "repeatable-read" isolation level

This program adds a determination function to detect test cases in MySQL environment. For now, "repeatable-read" and "serializable" isolation levels are supported. Other isolation levels and dbms environments will be added later.
The output result passes almost all test cases, except "rat_mda_step_rat_predicate_based_insert", because it uses "sum" function so an improved syntax analysis is required.

Tang-Kin-Long committed:
test: pass test case in Mysql at ALL isolation levels

This program adds two determination functions to detect test cases in MySQL environment. For now, ALL isolation levels in MySQL are supported. Other dbms environments will be added if needed.

The output result passes almost all test cases, except: 
SERIALIZABLE: "iat_dda_write_skew_predicate_based-overdraft_protection";
REPEATABLE-READ: "rat_mda_step_rat_predicate_based_insert";
READ-COMMITTED: None;
READ-UNCOMMITTED: "iat_dda_write_skew_predicate_based-intersecting_data";

These cases use irregular expression like "sum" function, self-defined table name or attribute query, so the more complex syntax analysis is required.
